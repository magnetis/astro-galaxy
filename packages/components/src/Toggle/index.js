import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { space } from 'styled-system';

const TOGGLE_BASE_WIDTH = '50px';
const TOGGLE_BASE_HEIGHT = '30px';
const TOGGLE_CIRCLE_SIZE = '28px';

const ToggleButton = styled.button`
  color: ${props => props.theme.colors.moon900};
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
  font: ${props => props.theme.fonts.secondary};

  /**
    Allow us to use custom margin or padding via props if needed.
    For more info check styled-system docs: https://styled-system.com/theme-specification
  */
  ${space};

  span {
    position: relative;
    display: inline-block;
    vertical-align: middle;
    margin-left: 8px;

    &:before {
      display: block;
      width: ${TOGGLE_BASE_WIDTH};
      height: ${TOGGLE_BASE_HEIGHT};
      content: '';
      background-color: ${props => props.theme.colors.moon400};
      border-radius: ${TOGGLE_CIRCLE_SIZE};
      transition: background-color 0.3s ease;
    }

    &:after {
      position: absolute;
      top: 1px;
      left: 1px;
      display: block;
      width: ${TOGGLE_CIRCLE_SIZE};
      height: ${TOGGLE_CIRCLE_SIZE};
      visibility: visible;
      content: '';
      background-color: ${props => props.theme.colors.space100};
      border-radius: ${TOGGLE_CIRCLE_SIZE};
      transition: left 0.3s ease;
    }
  }

  &[aria-pressed] {
    :not(:disabled):hover {
      & span {
        &:before {
          background-color: ${props => props.theme.colors.moon300};
        }
      }
    }
  }

  &[aria-pressed='true'] {
    & span {
      &:before {
        background-color: ${props => props.theme.colors.uranus500};
      }

      &:after {
        left: 21px;
      }
    }

    :not(:disabled):hover {
      & span {
        &:before {
          background-color: ${props => props.theme.colors.uranus400};
        }
      }
    }
  }

  :disabled {
    color: ${props => props.theme.colors.moon200};
    cursor: not-allowed;

    & span {
      &:before {
        background-color: ${props => props.theme.colors.moon200};
      }
    }
  }

  :focus {
    outline: none;

    & span {
      &:before {
        box-shadow: 0 0 0 3px rgba(21, 156, 228, 0.4);
      }
    }
  }
`;

ToggleButton.displayName = 'ToggleButton';

const Toggle = ({ label, isToggled, ...rest }) => {
  return (
    <ToggleButton type="button" aria-pressed={isToggled} {...rest}>
      {label}
      <span aria-hidden="true" />
    </ToggleButton>
  );
};

Toggle.propTypes = {
  label: PropTypes.string.isRequired,
  isToggled: PropTypes.bool.isRequired,
  onClick: PropTypes.func,
};

Toggle.defaultProps = {
  isToggled: false,
};

export default Toggle;
