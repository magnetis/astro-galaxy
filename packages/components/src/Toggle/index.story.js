import { storiesOf } from '@storybook/react';
import React, { useState } from 'react';
import Toggle from './index.js';

const ToggleSample = ({ disabled }) => {
  const [isToggled, setToggle] = useState(false);

  return (
    <>
      <Toggle
        label="default"
        isToggled={isToggled}
        onClick={() => setToggle(!isToggled)}
        disabled={disabled}
      />
    </>
  );
};

storiesOf('controls & toggles|toggles', module)
  .add('default', () => <ToggleSample />)
  .add('disabled', () => <ToggleSample disabled />);
