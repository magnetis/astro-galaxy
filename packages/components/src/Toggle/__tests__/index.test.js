import React from 'react';
import Toggle from '../index';

describe('Toggle', () => {
  it('should render a toggle button', () => {
    const json = rendererCreateWithTheme(<Toggle label="resgate total" />).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should render a toggled toggle button', () => {
    const json = rendererCreateWithTheme(
      <Toggle isToggled={true} label="resgate total" />
    ).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should render a disabled toggle button', () => {
    const json = rendererCreateWithTheme(<Toggle disabled label="resgate total" />).toJSON();

    expect(json).toMatchSnapshot();
  });
});
