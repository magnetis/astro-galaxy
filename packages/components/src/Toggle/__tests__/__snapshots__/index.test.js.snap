// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render a disabled toggle button 1`] = `
.c0 {
  color: #32383c;
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
  font: normal normal 400 normal 16px/1.5 Lato,sans-serif;
}

.c0 span {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  margin-left: 8px;
}

.c0 span:before {
  display: block;
  width: 50px;
  height: 30px;
  content: '';
  background-color: #8296a4;
  border-radius: 28px;
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
}

.c0 span:after {
  position: absolute;
  top: 1px;
  left: 1px;
  display: block;
  width: 28px;
  height: 28px;
  visibility: visible;
  content: '';
  background-color: #ffffff;
  border-radius: 28px;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.c0[aria-pressed]:not(:disabled):hover span:before {
  background-color: #afbec9;
}

.c0[aria-pressed='true'] span:before {
  background-color: #159ce4;
}

.c0[aria-pressed='true'] span:after {
  left: 21px;
}

.c0[aria-pressed='true']:not(:disabled):hover span:before {
  background-color: #3db5eb;
}

.c0:disabled {
  color: #e9eef2;
  cursor: not-allowed;
}

.c0:disabled span:before {
  background-color: #e9eef2;
}

.c0:focus {
  outline: none;
}

.c0:focus span:before {
  box-shadow: 0 0 0 3px rgba(21,156,228,0.4);
}

<button
  aria-pressed={false}
  className="c0"
  disabled={true}
  type="button"
>
  resgate total
  <span
    aria-hidden="true"
  />
</button>
`;

exports[`Toggle should render a toggle button 1`] = `
.c0 {
  color: #32383c;
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
  font: normal normal 400 normal 16px/1.5 Lato,sans-serif;
}

.c0 span {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  margin-left: 8px;
}

.c0 span:before {
  display: block;
  width: 50px;
  height: 30px;
  content: '';
  background-color: #8296a4;
  border-radius: 28px;
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
}

.c0 span:after {
  position: absolute;
  top: 1px;
  left: 1px;
  display: block;
  width: 28px;
  height: 28px;
  visibility: visible;
  content: '';
  background-color: #ffffff;
  border-radius: 28px;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.c0[aria-pressed]:not(:disabled):hover span:before {
  background-color: #afbec9;
}

.c0[aria-pressed='true'] span:before {
  background-color: #159ce4;
}

.c0[aria-pressed='true'] span:after {
  left: 21px;
}

.c0[aria-pressed='true']:not(:disabled):hover span:before {
  background-color: #3db5eb;
}

.c0:disabled {
  color: #e9eef2;
  cursor: not-allowed;
}

.c0:disabled span:before {
  background-color: #e9eef2;
}

.c0:focus {
  outline: none;
}

.c0:focus span:before {
  box-shadow: 0 0 0 3px rgba(21,156,228,0.4);
}

<button
  aria-pressed={false}
  className="c0"
  type="button"
>
  resgate total
  <span
    aria-hidden="true"
  />
</button>
`;

exports[`Toggle should render a toggled toggle button 1`] = `
.c0 {
  color: #32383c;
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
  font: normal normal 400 normal 16px/1.5 Lato,sans-serif;
}

.c0 span {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  margin-left: 8px;
}

.c0 span:before {
  display: block;
  width: 50px;
  height: 30px;
  content: '';
  background-color: #8296a4;
  border-radius: 28px;
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
}

.c0 span:after {
  position: absolute;
  top: 1px;
  left: 1px;
  display: block;
  width: 28px;
  height: 28px;
  visibility: visible;
  content: '';
  background-color: #ffffff;
  border-radius: 28px;
  -webkit-transition: left 0.3s ease;
  transition: left 0.3s ease;
}

.c0[aria-pressed]:not(:disabled):hover span:before {
  background-color: #afbec9;
}

.c0[aria-pressed='true'] span:before {
  background-color: #159ce4;
}

.c0[aria-pressed='true'] span:after {
  left: 21px;
}

.c0[aria-pressed='true']:not(:disabled):hover span:before {
  background-color: #3db5eb;
}

.c0:disabled {
  color: #e9eef2;
  cursor: not-allowed;
}

.c0:disabled span:before {
  background-color: #e9eef2;
}

.c0:focus {
  outline: none;
}

.c0:focus span:before {
  box-shadow: 0 0 0 3px rgba(21,156,228,0.4);
}

<button
  aria-pressed={true}
  className="c0"
  type="button"
>
  resgate total
  <span
    aria-hidden="true"
  />
</button>
`;
